From ffd1f85f70551c897e6e62b89acf8b170c1c82ea Mon Sep 17 00:00:00 2001
From: "M. Vefa Bicakci" <vefa.bicakci@windriver.com>
Date: Thu, 31 Mar 2022 16:14:22 -0400
Subject: [PATCH] configure.ac: Detect rdma-core v36 and v37 too

This commit makes the configure.ac script detect rdma-core v36 and v37
as well. Of these, we are only interested in v37, because Mellanox's
OpenFabrics Enterprise Distribution is based on rdma-core v37.

Signed-off-by: M. Vefa Bicakci <vefa.bicakci@windriver.com>
---
 configure.ac | 76 +++++++++++++++++++++++++++++++++++++++++++++++++++-
 1 file changed, 75 insertions(+), 1 deletion(-)

diff --git a/configure.ac b/configure.ac
index 9e4e980886d4..3ece437e664a 100644
--- a/configure.ac
+++ b/configure.ac
@@ -68,6 +68,10 @@ m4_define([rcore_version],[m4_esyscmd(rc=$(grep -o 		\
 	_ibv_device_ops verbsincpath/include/infiniband/verbs.h |	\
 	tail -1);						\
 	if test ! -z "$rc"; then				\
+		rcverv37=$(grep "IBV_WC_DRIVER3" \
+			   verbsincpath/include/infiniband/verbs.h|tail -1);	\
+		rcverv36=$(grep "ibv_query_qp_data_in_order" \
+			   verbsincpath/include/infiniband/verbs.h|tail -1);	\
 		rcverv35=$(grep "ibv_is_fork_initialized" \
 			   verbsincpath/include/infiniband/verbs.h|tail -1);	\
 		rcverv34=$(grep "ibv_reg_dmabuf_mr" \
@@ -90,7 +94,11 @@ m4_define([rcore_version],[m4_esyscmd(rc=$(grep -o 		\
 			verbsincpath/include/infiniband/verbs.h|tail -1);	\
 		rcverv17=$(grep -o ibv_alloc_td			\
 			verbsincpath/include/infiniband/verbs.h|tail -1);	\
-		if test ! -z "$rcverv35"; then			\
+		if test ! -z "$rcverv37"; then			\
+			echo -n "v37";				\
+		elif test ! -z "$rcverv36"; then			\
+			echo -n "v36";				\
+		elif test ! -z "$rcverv35"; then			\
 			echo -n "v35";				\
 		elif test ! -z "$rcverv34"; then			\
 			echo -n "v34";				\
@@ -175,6 +183,72 @@ if test "$rc" != ""; then
 	AC_DEFINE([HAVE_IBV_CMD_MODIFY_QP_EX],[1],[ibv_cmd_modify_qp_ex supported])
 	RDMA_CORE_HEADERS=`echo -n __include_path__`
 	case  "$(echo -n rcore_version)" in
+	v37)
+		AC_MSG_RESULT(yes using v37)
+		AC_DEFINE([IBV_WC_DRIVER3_IN_IBV_WC_OPCODE],[1],
+			  [Rdma-Core Package has IBV_WC_DRIVER3 in ibv_wc_opcode enum])
+		AC_DEFINE([IBV_WC_DRIVER2_IN_IBV_WC_OPCODE],[1],
+			  [Rdma-Core Package has IBV_WC_DRIVER2 in ibv_wc_opcode enum])
+		AC_DEFINE([IBV_QUERY_QP_DATA_IN_ORDER_IN_CONTEXT_OPS],[1],
+			  [Rdma-Core Package has query_qp_data_in_order in verbs_context_ops])
+		AC_DEFINE([VERBS_ONLY_QUERY_DEVICE_EX_DEFINED],[1],
+			  [Rdma-Core Package has only query_device_ex])
+		AC_DEFINE([IBV_CMD_MODIFY_QP_EX_HAS_7_ARG],[1],
+			  [Rdma-Core Package ibv_cmd_modify_qp_ex has 7 arguments])
+		AC_DEFINE([IBV_FREE_CONTEXT_IN_CONTEXT_OPS],[1],
+			  [Rdma-Core Package has free context in ibv_context_ops])
+		AC_DEFINE([REG_MR_VERB_HAS_5_ARG],[1],
+			  [Rdma-Core Package reg_mr has 5 argument])
+		AC_DEFINE([IBV_CMD_ALLOC_MW_HAS_1_ARG],[1],
+			  [Rdma-Core Package ibv_cmd_alloc_mw has single argument])
+		AC_DEFINE([PROVIDER_DRIVER_HAS_2_ARGS],[1],
+			  [Rdma-Core Package PROVIDER_DRIVER macro has 2 arguments])
+		AC_DEFINE([ALLOC_CONTEXT_HAS_PRIVATE_DATA],[1],
+			  [Rdma-Core Package alloc_context has private data])
+		AC_DEFINE([VERBS_MR_DEFINED],[1],
+			  [Rdma-Core Package has verbs_mr defined])
+		AC_DEFINE([VERBS_INIT_AND_ALLOC_CONTEXT_HAS_5_ARG],[1],
+			  [Rdma-Core Package has 5 Arg for verbs_init_and_alloc_context macro])
+		AC_DEFINE([RCP_HAS_PROVIDER_DRIVER],[1],
+			  [Rdma-Core Package has PROVIDER_DRIVER macro])
+		AC_DEFINE([RCP_USE_IB_UVERBS],[1],
+			[Rdma-Core Package uses IB user verbs API])
+		AC_DEFINE([RCP_USE_ALLOC_CONTEXT],[1],
+			[Rdma-Core Package uses alloc_context instead of init_context])
+		rc=`grep -o ibv_read_sysfs_file src/rc-compat/v37/driver.h|tail -1`
+	;;
+	v36)
+		AC_MSG_RESULT(yes using v36)
+		AC_DEFINE([IBV_WC_DRIVER2_IN_IBV_WC_OPCODE],[1],
+			  [Rdma-Core Package has IBV_WC_DRIVER2 in ibv_wc_opcode enum])
+		AC_DEFINE([IBV_QUERY_QP_DATA_IN_ORDER_IN_CONTEXT_OPS],[1],
+			  [Rdma-Core Package has query_qp_data_in_order in verbs_context_ops])
+		AC_DEFINE([VERBS_ONLY_QUERY_DEVICE_EX_DEFINED],[1],
+			  [Rdma-Core Package has only query_device_ex])
+		AC_DEFINE([IBV_CMD_MODIFY_QP_EX_HAS_7_ARG],[1],
+			  [Rdma-Core Package ibv_cmd_modify_qp_ex has 7 arguments])
+		AC_DEFINE([IBV_FREE_CONTEXT_IN_CONTEXT_OPS],[1],
+			  [Rdma-Core Package has free context in ibv_context_ops])
+		AC_DEFINE([REG_MR_VERB_HAS_5_ARG],[1],
+			  [Rdma-Core Package reg_mr has 5 argument])
+		AC_DEFINE([IBV_CMD_ALLOC_MW_HAS_1_ARG],[1],
+			  [Rdma-Core Package ibv_cmd_alloc_mw has single argument])
+		AC_DEFINE([PROVIDER_DRIVER_HAS_2_ARGS],[1],
+			  [Rdma-Core Package PROVIDER_DRIVER macro has 2 arguments])
+		AC_DEFINE([ALLOC_CONTEXT_HAS_PRIVATE_DATA],[1],
+			  [Rdma-Core Package alloc_context has private data])
+		AC_DEFINE([VERBS_MR_DEFINED],[1],
+			  [Rdma-Core Package has verbs_mr defined])
+		AC_DEFINE([VERBS_INIT_AND_ALLOC_CONTEXT_HAS_5_ARG],[1],
+			  [Rdma-Core Package has 5 Arg for verbs_init_and_alloc_context macro])
+		AC_DEFINE([RCP_HAS_PROVIDER_DRIVER],[1],
+			  [Rdma-Core Package has PROVIDER_DRIVER macro])
+		AC_DEFINE([RCP_USE_IB_UVERBS],[1],
+			[Rdma-Core Package uses IB user verbs API])
+		AC_DEFINE([RCP_USE_ALLOC_CONTEXT],[1],
+			[Rdma-Core Package uses alloc_context instead of init_context])
+		rc=`grep -o ibv_read_sysfs_file src/rc-compat/v36/driver.h|tail -1`
+	;;
 	v35)
 		AC_MSG_RESULT(yes using v35)
 		AC_DEFINE([VERBS_ONLY_QUERY_DEVICE_EX_DEFINED],[1],
-- 
2.29.2

